client/trino-client/src/main/java/io/trino/client/auth/external/CashedToken.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.client.auth.external;
+
+import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
+
+/**
+ * This KnownToken instance forces all Connections to reuse same token.
+ * Every time an existing token is considered to be invalid each Connection
+ * will try to obtain a new token, but only the first one will actually do the job,
+ * where every other connection will be waiting on readLock
+ * until obtaining new token finishes.
+ *
+ * In general the game is to reuse same token and obtain it only once, no matter how
+ * many Connections will be actively using it. It's very important as obtaining the new token
+ * will take minutes, as it mostly requires user thinking time.
+ */
+class CashedToken
+        implements KnownToken
+{
+    private static final ReadWriteLock lock = new ReentrantReadWriteLock();
```

Comment:
Can we have it as regular singleton, where you have single `static instance` and other fields are not static.